samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,379 @@
+    if (window.getAccumulationMode() == AccumulationMode.DISCARDING) {
+      store.put(windowKey, null);
+    }
+
+    // Cancel all early triggers too when the default trigger fires. Also, clean all state for the key.
+    // note: We don't handle late arrivals yet, So, all arrivals are either early or on-time.
+    if (triggerKey.getType() == TriggerType.DEFAULT) {
+      cancelTrigger(triggerKey, true);
+      cancelTrigger(new TriggerKey(TriggerType.EARLY, triggerKey.getKey()), true);
+      WindowKey<K> key = triggerKey.key;
+      store.delete(key);
+    }
+
+    // Cancel non-repeating early triggers.
+    if (triggerKey.getType() == TriggerType.EARLY && !wrapper.isRepeating()) {
+      cancelTrigger(triggerKey, false);
+    }
+  }
+
+  /**
+   * Computes the pane output corresponding to a {@link TriggerKey} that fired.
+   */
+  private WindowPane<K, WV> computePaneOutput(TriggerKey<K> triggerKey, WindowState<WV> state) {
+    WindowKey<K> windowKey = triggerKey.key;
+    WV windowVal = state.getWindowValue();
+
+    //For session windows, we will create a new window key by using the time of the first message in the window as
+    //the paneId.
+    if (window.getWindowType() == WindowType.SESSION) {
+      windowKey = new WindowKey<>(windowKey.getKey(), Long.toString(state.getEarliestTime()));
+    }
+
+    //Make a defensive copy so that we are immune to further mutations on the collection
+    if (windowVal instanceof Collection) {
+      windowVal = (WV) new ArrayList<>((Collection<WV>) windowVal);
+    }
+
+    WindowPane<K, WV> paneOutput = new WindowPane<>(windowKey, windowVal, window.getAccumulationMode(), triggerKey.getType());
+    return paneOutput;
+  }
+
+  /**
+   * Cancels the firing of the {@link TriggerImpl} identified by this {@link TriggerKey} and optionally removes it.
+   */
+  private void cancelTrigger(TriggerKey<K> triggerKey, boolean shouldRemove) {
+    TriggerImplState wrapper = triggers.get(triggerKey);
+    if (wrapper != null) {
+      wrapper.cancel();
+    }
+    if (shouldRemove && triggerKey != null) {
+      triggers.remove(triggerKey);
+    }
+  }
+
+  /**
+   * Implementation of the {@link TriggerContext} that allows {@link TriggerImpl}s to schedule and cancel callbacks.
+   */
+  public class TriggerContextImpl implements TriggerContext {
+    private final TriggerKey<K> triggerKey;
+
+    public TriggerContextImpl(TriggerKey<K> triggerKey) {
+      this.triggerKey = triggerKey;
+    }
+
+    @Override
+    public Cancellable scheduleCallback(Runnable runnable, long callbackTimeMs) {
+      TriggerCallbackState timerState = new TriggerCallbackState(triggerKey, runnable, callbackTimeMs);
+      pendingCallbacks.add(timerState);
+      return timerState;
+    }
+  }
+
+  /**
+   * State corresponding to pending timer callbacks scheduled by various {@link TriggerImpl}s.
+   */
+  private class TriggerCallbackState implements Comparable<TriggerCallbackState>, Cancellable {
```

Comment:
Actually, one other way to do this could be:

`TriggerCallbackState(TriggerKey<WK> triggerKey, Runnable callback, long scheduledTimeMs)` to pass in the `WindowOperatorImpl` into the `TriggerCallbackState`'s constructor.

That would make `TriggerCallbackState<M, WK, WV>` be typed as well. I'd be nice to not have this verbosity for storing something simple as a combination of a key, and a Runnable :-)